feat(map_projection_loader): add map_projection_loader

kminoda opened the pull request:
## Description

This PR adds `map_projection_loader` that loads the map projection info and publishes it as `/map/map_projector_type`. The topic will be subscribed by `/map/lanelet2_map_loader`. 

Importantly, **the interface for map files will change with this PR**. Autoware will require `map_projector_info.yaml` as an additional file that stores the projection info. To maintain backward compatibility, we also support maps without the yaml file, but note that this will evoke a deprecated warning. For details, please refer to [the readme of `map_projection_loader` which is added in this PR](https://github.com/kminoda/autoware.universe/blob/feat/add_map_projection_loader/map/map_projection_loader/README.md). 


<!-- Write a brief description of this PR. -->

## Related links

Should be merged with:
- https://github.com/autowarefoundation/autoware_launch/pull/483

Related links:
- https://github.com/orgs/autowarefoundation/discussions/3582
- [TIER IV INTERNAL LINK](https://tier4.atlassian.net/wiki/spaces/LM/pages/2799341381/Super+Large+Size+Map+Design+Doc)
<!-- Write the links related to this PR. Private links should be clearly marked as private, for example, '[FOO COMPANY INTERNAL LINK](https://example.com)'. -->

## Tests performed

### Unit test & Integration test
I have added some unit test and integration test under `map/map_projection_loader/test`.

### Testing with logging_simulator
I also tested the following 
- [backward-compatibility test] Confirmed that localization works with `sample-map-rosbag` provided in Autoware tutorial.
- [MGRS test from YAML] Prepare the following `map_projector_info.yaml` in `sample-map-rosbag` directory, launch logging_simulator, and confirmed that it works.

`map_projector_info.yaml` for sample-map-rosbag:
```yaml
type: MGRS
mgrs_grid: 54SUE
```

### Testing with planning_simulator
I also tested planning_simulator with `sample-map-planning` provided in Autoware tutorial.
<!-- Describe how you have tested this PR. -->

## Notes for reviewers
Transverse Mercator projection is not yet supported only with this PR, but we also need to add https://github.com/autowarefoundation/autoware.universe/pull/3984. I split PRs into two to maintain each PR as small as possible, but let me know if it's better to merge these at once. (Currently, I am considering merging this PR first and then open the PR)

<!-- Write additional information if necessary. It should be written if there are related PRs that should be merged at the same time. -->

## Interface changes
Map files requirement will change, although backward compatibility is also supported
<!-- Describe any changed interfaces, such as topics, services, or parameters. -->

## Effects on system behavior
None except the above interface change.
<!-- Describe how this PR affects the system behavior. -->

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

kminoda commented on map/map_projection_loader/src/map_projection_loader.cpp:
```diff
@@ -0,0 +1,59 @@
+    RCLCPP_WARN(
+      this->get_logger(),
+      "DEPRECATED WARNING: Loading map projection info from lanelet2 map may soon be deleted. "
+      "Please use map_projector_info.yaml instead.");
```
Maybe better to put URL here

kminoda replied:
https://github.com/autowarefoundation/autoware.universe/pull/3986/commits/6817598bb8351cfc3102bb16df5bd053ac465cd3

mitsudome-r commented:
I have discussed with @kminoda -san online. Here are some requests:
- Please update Autoware Documentation to explain about projection loader architecture
- split PRs for Transverse Mercator type support
- possibly improve backward compatibility for local projection maps
- update MGRSProjector in lanelet2_extension to use MGRS grid specified in the map_projection_loader (after this PR)

kminoda commented:
@mitsudome-r Thank you for the memo!
Open the PR again.
- I created a PR for Autoware Documentation as well: https://github.com/autowarefoundation/autoware-documentation/pull/431/
- I split PRs for Transverse Mercator type support
- I would NOT implement backward compatibility for local lanelet2, since it may make the API too complicated (taking lat lon from lanelet2 seems to be difficult, and using `mgrs_code` tag may make the things too complicated if we take into account customized lanelet2 by various users)
  - Countermeasures: I'd like to post a warning description right after this PR is finished. 


Discussion draft:
--
We have changed a map interface. Now, Autoware would require `map_projector_info.yaml` as well as `pointcloud_map.pcd` file(s) and `lanelet2_map.osm` file. The background and the motivation behind this change is [Discussion](https://github.com/orgs/autowarefoundation/discussions/3582). 

As a backward compatibility, if `map_projector_info.yaml` does not exist, Autoware will assume that it's loaded in MGRS coordinate, and the MGRS grid is determined by the coordinates of `lanelet2_map.osm`. Thus, those who are using local coordinate will be affected, so please kindly add `map_projector_info.yaml` in your map directory.

If you have any troubles related to maps (e.g. visualization on RViz), please go through [Map component design doc in Autoware Documentation](https://autowarefoundation.github.io/autoware-documentation/main/design/autoware-architecture/map/). In addition, the readme of [`map_loader`](https://github.com/autowarefoundation/autoware.universe/blob/main/map/map_loader/README.md) and [`map_projection_loader`](https://github.com/autowarefoundation/autoware.universe/blob/main/map/map_loader/README.md) may also be helpful.

Related links
- [Discussion](https://github.com/orgs/autowarefoundation/discussions/3582)
- [One of the PRs](https://github.com/autowarefoundation/autoware.universe/pull/3986)

YamatoAndo commented on map/map_projection_loader/test/data/projection_info_local.yaml:
```diff
@@ -0,0 +1,6 @@
+type: MGRS
```
```suggestion
type: Local
```

kminoda replied:
Fixed in https://github.com/autowarefoundation/autoware.universe/pull/3986/commits/b358ab93eb297997ec416269b7738fac822a4733

mitsudome-r commented on map/map_loader/README.md:
```diff
@@ -154,10 +159,11 @@ The node projects lan/lon coordinates into MGRS coordinates.
 ### Feature
 
-lanelet2_map_visualization visualizes autoware_auto_mapping_msgs/HADMapBin messages into visualization_msgs/MarkerArray. There are 3 types of map can be loaded in autoware. Please make sure you selected the correct lanelet2_map_projector_type when you launch this package.
+lanelet2_map_visualization visualizes autoware_auto_mapping_msgs/HADMapBin messages into visualization_msgs/MarkerArray. There are four types of map can be loaded in autoware. Please make sure you selected the correct lanelet2_map_projector_type when you launch this package.
 
 - MGRS
 - UTM
+- TransverseMercator
 - local
```
This isn't your fault, but description about projection should go under explanation for lanelet2_map_loader [at L140](https://github.com/autowarefoundation/autoware.universe/blob/cd09c426cb1cbb2c80d7ab56e7408930b71c5438/map/map_loader/README.md?plain=1#L140), and not under visualization because it doesn't do any projection in visualization node.

```suggestion
lanelet2_map_visualization visualizes autoware_auto_mapping_msgs/HADMapBin messages into visualization_msgs/MarkerArray. 
```

kminoda replied:
Fixed in https://github.com/autowarefoundation/autoware.universe/pull/3986/commits/4b8ddbbb0bd9f0870cd822993b0d3fc44c866fbe

mitsudome-r commented on map/map_loader/README.md:
```diff
@@ -138,12 +139,16 @@ Please see [the description of `GetSelectedPointCloudMap.srv`](https://github.co
 
 lanelet2_map_loader loads Lanelet2 file and publishes the map data as autoware_auto_mapping_msgs/HADMapBin message.
-The node projects lan/lon coordinates into MGRS coordinates.
+The node projects lan/lon coordinates into arbitrary coordinates defined in `/map/map_projector_type` from `map_projection_loader`.
 
```
```suggestion
The node projects lan/lon coordinates into arbitrary coordinates defined in `/map/map_projector_type` from `map_projection_loader`.
The node supports the following four types of coordinate systems:

- MGRS
- UTM
- TransverseMercator
- local
```

kminoda replied:
Fixed in https://github.com/autowarefoundation/autoware.universe/pull/3986/commits/4b8ddbbb0bd9f0870cd822993b0d3fc44c866fbe

mitsudome-r commented on map/map_loader/README.md:
```diff
@@ -174,10 +180,7 @@ lanelet2_map_visualization visualizes autoware_auto_mapping_msgs/HADMapBin messa
-| longitude                   | double      | Longitude of map_origin, only using in UTM map projector     | 0.0           |
-| center_line_resolution      | double      | Define the resolution of the lanelet center line             | 5.0           |
-| lanelet2_map_path           | std::string | The lanelet2 map path                                        | None          |
+| Name                   | Type        | Description                                      | Default value |
+| :--------------------- | :---------- | :----------------------------------------------- | :------------ |
+| center_line_resolution | double      | Define the resolution of the lanelet center line | 5.0           |
+| lanelet2_map_path      | std::string | The lanelet2 map path                            | None          |
```
These should be moved under `lanelet2_map_loader`

kminoda replied:
Fixed in https://github.com/autowarefoundation/autoware.universe/pull/3986/commits/4b8ddbbb0bd9f0870cd822993b0d3fc44c866fbe

YamatoAndo commented:
@kminoda please fix build error

> __w/autoware.universe/autoware.universe/map/map_projection_loader/src/map_projection_loader.cpp:35:5: error: continue statement not within a loop
>    35 |     continue;  // do nothing

kminoda commented:
@YamatoAndo Thank you :pray: 
Fixed https://github.com/autowarefoundation/autoware.universe/pull/3986/commits/0702af6c7688c4b25c4c64da7d329e60d2a3fa4a

YamatoAndo commented on launch/tier4_map_launch/launch/map.launch.py:
```diff
@@ -57,7 +62,10 @@ def launch_setup(context, *args, **kwargs):
-        remappings=[("output/lanelet2_map", "vector_map")],
+        remappings=[
+            ("output/lanelet2_map", "vector_map"),
+            ("input/map_projector_info", "map_projector_type"),
```
@kminoda 
Why did you use "input/map_projector_info" instead of "input/map_projector_type"?
I believe using the same name "map_projector_info" as the message type would be more intuitive.

kminoda replied:
Thank you! I understood that it's rather a question to the current implementation of Autoware.
@mitsudome-r @tkhmy Do you know anything about this?

mitsudome-r replied:
I don't think there are any reasons. I also think projector info is better as well.

tkhmy replied:
I think initially I just publish mgrs, so I use type.
Then after that the topic also publish information, so change to info , so I think /map/map_projector_info will be a better naming

kminoda replied:
Thank you for the info.
How about this?
- This PR will be merged as it is
- The topic name will be changed in another PR after this

YamatoAndo replied:
OK!

YamatoAndo commented on map/map_projection_loader/package.xml:
```diff
@@ -0,0 +1,27 @@
+  <name>map_projection_loader</name>
+  <version>0.1.0</version>
+  <description>map_projection_loader package as a ROS 2 node</description>
+  <maintainer email="koji.minoda@tier4.jp">Koji Minoda</maintainer>
```
@kminoda Can another person, in addition to you, be assigned as a maintainer?

kminoda replied:
Fixed: https://github.com/autowarefoundation/autoware.universe/pull/3986/commits/378d9dc8ae70613bd382f248e88f04a56a1a4197

YamatoAndo commented:
I have confirmed that Autoware works well with the logging_simulator/planning_simulator.
I have also confirmed that the `/map/map_projector_type` topic is being output as intended.

kminoda commented:
@YamatoAndo @mitsudome-r Hi, I've reflected all the comments! Would you mind checking the PR and approve if it looks OK?
